Timeout on client

edouardproust opened the pull request:
I hope i didn't break too many things :/
I added a Client class + the timeout on client sockets. It disconnect them simply, no error response sent.

Wow this PR is almost 2 times more lines than what Daniel ever did on this project, incredible.

edouardproust committed:
update

edouardproust committed:
fixed py script

edouardproust committed:
todos for canonical clesses check

Skoteini-42 commented:
I think the whole Client implementation is a smart addtion and makes our server even more complete! I mean, it's how Nginx works as well as I tried to read it's open source code and I really REALLY like it. A few thoughts I have and noticed :

1) I think time-out for 10 seconds in case of inactivity is too short, because it seems to me it breaks the keep-alive concept. So maybe we can have it for security just like you said, but a bit longer? Maybe 75 seconds, I timed nginx, and it times out then :) Nginx uses different timeouts for different situations, so maybe we can also have a `keep-alive_timeout =75` and a different timeout for hanging requests with 10 sec, what do you think ?

2) You do this check here inside R`equestParser::isRawRequestComplete`

```
		if (bodyLength < contentLength)
			return false;
```
I understand you put that for this eg we were talking about yesterday, : `printf "GET / HTTP/1.1\r\nHost: localhost\r\nContent-length: 100\r\n\r\n"hello | nc localhost 8080`, Nginx returns 200 AND times out after 5 seconds. And I also commented (while the fight was happening in slack, so maybe you didn't see that) that if you put a very large number in content-length, it returns 413 instead, AND times out after 5 seconds eg : printf `"GET / HTTP/1.1\r\nHost: localhost\r\nContent-length: 10000000\r\n\r\n"hello | nc localhost 8080`, so that tells us that it's timing out according to the body max size it has (and that also tells us that it uses a different timeout for bodies...but mayeb we don't care about that, we can use 10 secs here too).
I don't think we're expected to match nginx's behaviour 100% everywhere, I am just commenting this in case you have another idea. what we're doing now is always timing out if `bodyLength < contentLength` without responses, maybe we want to keep it this way or not?

3) Inside `Client::continueResponseSend()` line 68 sends a position value isntead of fd : `Log::prod("ok", "Response sent to client (fd " + Log::hl(_responseSendPos) + ").");` So we get this in the screen for eg : `2025-12-09 13:31:55 [OK] Response sent to client (fd 12939).` Minor detail, but easy fix!

4) There is some bug, if I test Connection: close, eg : `printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\n\r\n" | nc localhost 8080`, not only it returns 400 while this should be 200, but it keeps sending the body multiple times, and not sure if the connection even closes..ouch

----------------- Other--------------------

Just a check from you too, are we allowed to do this inside function arguments? (I mean this : bool=false)

`void		_sendErrorResponse(std::string const&, int, CgiData const*, std::string const&, bool = false); `

edouardproust committed:
fixed loop if connection close. Added different client timeout for keep alive.

edouardproust commented:
You point 1.:

i added a different client timeouts.

In the list below i mean the following:
- "Complete raw-request" = fully valid (thanks to RequestParser rules) or with body longer than the Content-Length value
- "Incomplete raw-request" = no headers-body seperator "\r\n\r\n" or a body shorter than the Content-Length value

Cases:

1. Complete raw-request with "Connection: keep-alive" header **-> we SEND A RESPONSE and use KEEP-ALIVE TIMEOUT:**
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: keep-alive\r\n\r\n" | nc localhost 8080` -> 200 OK
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: keep-alive\r\nContent-Length:3\r\n\r\nhello world" | nc localhost 8080` -> 400 bad request

2. Incomplete raw-request (whatever the value of the "Connection:" header value) **-> we SEND NO RESPONSE and use CLOSE TIMEOUT.** 
`printf "" | nc localhost 8080` (no request)
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: close" | nc localhost 8080`  (missing separator)
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: keep-alive" | nc localhost 8080`  (missing separator)
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: keep-alive\r\nContent-length:10000\r\n\r\nHello" | nc localhost 8080` (shorter body than the value in Content-Length)

3. Complete raw-request with "Connection: close" header -> **we SEND RESPONSE and we USE NO TIMEOUT (instant close):**
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\n\r\n" | nc localhost 8080` -> 200 OK
`printf "GET / HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\nContent-Length:3\r\n\r\nhello world" | nc localhost 8080` -> 400 Bad request

edouardproust commented:
Point 2.:
I don't want to do the same behaviour as `nginx`  on this. I don't know how to check that the body is fully received if the Content-Length value is longer than the actual size of the body in the raw request., and i think we don't have to do it. Is that ok for you ?

Point 3.: fixed

Point 4.: fixed

Extra: I think we can, do you saw anywhere we cannot do it ? We could argue we are using a value inside a hpp. I am not 100% sure about that, could you ask in Slack please?

Extra2:
I made conditional the check of `\n\n` inside `utils::headersBodySeparatorPos()` (now acception a bool param). And i set it to false fro Request and true for Response (the CGI can use \n\n in its raw response). The removal of this extra `find` should increase a bit the speed of the reception of the raw request from client for big requests (100MB for ex.)

edouardproust committed:
fixed points 2 3 4 of ava's comment in PR

Skoteini-42 commented:
Everything seems to be ok with the Client implementation + your recent fixes on this from today, it's really good!

Siege tests pass normally, I did the ones we have home.htm + more like these :
All show 100% available with no other issues or very delayed responses

siege -c 10 -t 30s http://localhost:8080/
siege -c 25 -t 30s http://localhost:8080/
siege -c 50 -t 30s http://localhost:8080/
siege -c 100 -t 30s http://localhost:8080/
siege -c 30 -t 30s --header="Connection: keep-alive" http://localhost:8080/
siege -c 30 -t 30s --header="Connection: close" http://localhost:8080/

Concerning the ubuntu_tester, we get a seg fault in the last test :

`Test multiple workers(20) doing multiple times(5): Post on /directory/youpi.bla with size 100000000`

Before this branch the only difference is that this test took forever for me ( I waited 40 miutes, i think more like an hour) and it was still passing data) but I closed it myself after this as I found it ridiculous. Now I see a seg fault and quite fast, that's the difference. Do we care ????


```
2025-12-10 19:59:30 [WARNING] CGI timeout: process 101591 exceeded 20 seconds (closed after 20 sec).
2025-12-10 19:59:30 [CLOSE] Killed CGI process 101591
2025-12-10 19:59:30 [ERROR] CGI timeout error
make: *** [Makefile:147: test_42] Segmentation fault (core dumped
```


Valgrind it's huge and repeatable so I cut it to the main issues :


==139265== Memcheck, a memory error detector
==139265== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==139265== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==139265== Command: ./webserv_42 tests/config/replaced/ubuntu_tester.config
==139265== Parent PID: 139264
==139265== 
==139265== Invalid read of size 8
==139265==    at 0x49CAD14: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::size() const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.33)
==139265==    by 0x110E72: bool std::operator==<char, std::char_traits<char>, std::allocator<char> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x11D20D: bool std::operator!=<char, std::char_traits<char>, std::allocator<char> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x15465F: StaticHandler::builtinError(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x154298: StaticHandler::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Request const&, std::map<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<int>, std::allocator<std::pair<int const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x15D946: CgiHandler::_sendErrorResponse(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, CgiData const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x15C571: CgiHandler::_handleTimeout(CgiContext*, long) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x15B2AC: CgiHandler::checkCompletion() (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x12E9B8: Network::startServers() (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x10D588: main (in /home/fpapadak/webserv/webserv_42)
==139265==  Address 0x51748d0 is 112 bytes inside a block of size 616 free'd
==139265==    at 0x484A164: operator delete(void*) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==139265==    by 0x13231C: Network::_disconnectClient(Client*) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x130953: Network::_checkClientsInactivity() (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x12E9E1: Network::startServers() (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x10D588: main (in /home/fpapadak/webserv/webserv_42)
==139265==  Block was alloc'd at
==139265==    at 0x4846FA3: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==139265==    by 0x131FD2: Network::_acceptNewClient(int) (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x12E660: Network::startServers() (in /home/fpapadak/webserv/webserv_42)
==139265==    by 0x10D588: main (in /home/fpapadak/webserv/webserv_42)


(after 1 hour it's still running so I am going to stop it, but at the end it returns a seg fault like I posted above in teh comments)

edouardproust committed:
fixed ?

edouardproust committed:
fixed ?

Skoteini-42 commented:
Warnings that appear occassionally :

2025-12-11 16:23:03 [CGI] Queued 330 bytes to fd 27
2025-12-11 16:23:03 [SETUP] Epoll MOD (Sending CGI response) successful for fd 27.
2025-12-11 16:23:03 [SETUP] Epoll DEL (CGI cleanup outRead) successful for fd 33.
2025-12-11 16:23:03 [SETUP] Epoll DEL (CGI cleanup errRead) successful for fd 35.
2025-12-11 16:23:03 [SETUP] Epoll DEL (CGI cleanup inWrite) successful for fd 32.
2025-12-11 16:23:03 [CLOSE] Cleaned up pipes for CGI process 5179
2025-12-11 16:23:03 [WARNING] CGI timeout: process 5212 exceeded 20 seconds (closed after 21 sec).
2025-12-11 16:23:03 [CLOSE] Killed CGI process 5212
2025-12-11 16:23:03 [ERROR] CGI timeout error
2025-12-11 16:23:03 [CGI] Queued 330 bytes to fd 28

Skoteini-42 reviewed (approved):
PR approved!!! Finally getting rid of the tester......
